plutus-tx/src/PlutusTx/DataList.hs

```diff
@@ -0,0 +1,71 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE OverloadedLists #-}
+{-# LANGUAGE TypeFamilies    #-}
+{-# LANGUAGE ViewPatterns    #-}
+
+module PlutusTx.DataList (
+
+) where
+
+import PlutusTx.Builtins qualified as P
+import PlutusTx.Builtins.Internal qualified as BI
+import PlutusTx.IsData qualified as P
+import PlutusTx.Prelude hiding (map)
+
+import GHC.Exts qualified as H
+
+newtype List a = List P.BuiltinData
```

Comment:
You can define a regular list, wrapped in `asData`:

```haskell
AsData.asData
  [d| data List a = Nil | Cons a (List a) |]
```

This should make defining the list operations much easier since you can use `Nil` and `Cons`.